File: binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans = Integer.MIN_VALUE;
    
    public int maxPathSum(TreeNode root) {
        if (root == null) {
            return 0;
        }
        
        maxPathToLeafSum(root);
        
        return ans;
    }
    
    // Return height
    private int maxPathToLeafSum(TreeNode root) {
        if (root == null) {
            return 0;
        }
        
        int leftMaxSum = Math.max(maxPathToLeafSum(root.left), 0);
        int rightMaxSum = Math.max(maxPathToLeafSum(root.right), 0);
        
        // Update ans
        ans = Math.max(ans, leftMaxSum + rightMaxSum + root.val);
        
        return Math.max(leftMaxSum, rightMaxSum) + root.val;
    }
}

File: binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        
        ans = float("-inf")
        
        def dfs(root):
            if root is None:
                return 0
            
            nonlocal ans
            
            left_sum = max(dfs(root.left), 0)
            right_sum = max(dfs(root.right), 0)
            
            ans = max(ans, left_sum + right_sum + root.val)
            
            # Max path sum from root towards leaf (might not include leaf)
            return max(left_sum, right_sum) + root.val
        
        dfs(root)
        
        return ans